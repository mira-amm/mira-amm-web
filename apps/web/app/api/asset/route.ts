/**
 * @api {get} /asset Get details of an asset by ID
 */
import {NextRequest, NextResponse} from "next/server";
import {GeckoTerminalQueryResponses} from "@/web/shared/types";
import {MainnetUrl} from "../../../../../libs/web/src/utils/constants";

// Handle GET requests for /api/asset
export async function GET(req: NextRequest) {
  // Extract the 'id' query parameter from the URL
  const url = new URL(req.url);
  const assetId = url.searchParams.get("id");
  // Return a 400 error if no 'id' is provided
  if (!assetId) {
    return NextResponse.json({error: "Asset ID is required"}, {status: 400});
  }

  try {
    const assetUrl = `${MainnetUrl}/assets/${assetId}`;

    // Fetch asset details
    const response = await fetch(assetUrl);

    // Handle non-OK HTTP responses
    if (!response.ok) {
      if (response.status === 404) {
        return NextResponse.json(
          {error: `Asset with ID: ${assetId} not found`},
          {status: 404},
        );
      }
      throw new Error(`Failed to fetch asset data: ${response.statusText}`);
    }

    const asset = await response.json();

    // Transforming to desired format for Gecko Terminal
    const transformedAsset: GeckoTerminalQueryResponses.Asset = {
      id: asset.assetId,
      name: asset.name,
      symbol: asset.symbol,
      decimals: asset.decimals,
      totalSupply: asset.totalSupply,
      circulatingSupply: asset.circulatingSupply,
    };

    const assetResponse: GeckoTerminalQueryResponses.AssetResponse = {
      asset: transformedAsset,
    };

    return NextResponse.json(assetResponse);
  } catch (error) {
    console.error("Error fetching asset data:", error);
    return NextResponse.json(
      {error: "Failed to fetch asset data"},
      {status: 500},
    );
  }
}
