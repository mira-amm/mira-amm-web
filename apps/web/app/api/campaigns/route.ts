/**
 * @api {get} /campaigns Get list of epochs and their campaigns
 */
import {SentioJSONCampaignService} from "@/src/models/campaigns/Campaign";
import {JSONEpochConfigService} from "@/src/models/campaigns/JSONEpochConfigService";
import {NextRequest, NextResponse} from "next/server";
import path from "path";

const CACHE_DURATION = 300;
const CACHE_STALE_WHILE_REVALIDATE = 150;

// Example query
//   --url 'http://localhost:3000/api/campaigns/'
// if you want to filter for poolIds, or epochNumbers, you can add them as query params
// They take multiple values, comma separated
// eg: epochNumbers=1,2,3&...
// eg: poolIds=286c479da40dc953bddc3bb4c453b608bba2e0ac483b077bd475174115395e6b-f8f8b6283d7fa5b672b530cbb84fcccb4ff8dc40f8176ef4544ddb1f1952ad07-false&...

const CACHE_HEADERS = {
  "Content-Type": "application/json",
  "Cache-Control": `public, max-age=${CACHE_DURATION}, stale-while-revalidate=${CACHE_STALE_WHILE_REVALIDATE}`,
};

const epochConfigService = new JSONEpochConfigService(
  path.join(process.cwd(), "../../libs/web/src", "models", "campaigns.json"),
);

const campaignService =
  process.env.SENTIO_API_KEY && process.env.SENTIO_API_URL
    ? new SentioJSONCampaignService(
        process.env.SENTIO_API_URL,
        process.env.SENTIO_API_KEY,
        epochConfigService,
      )
    : null;

export async function GET(request: NextRequest): Promise<NextResponse> {
  try {
    if (!campaignService) {
      throw new Error("No Sentio API key or URL provided");
    }

    const searchParams = request.nextUrl.searchParams;
    const epochNumbers = searchParams.get("epochNumbers");
    const poolIds = searchParams.get("poolIds");
    const includeAPR = searchParams.get("includeAPR") === "true";

    const campaigns = await campaignService.getCampaigns({
      epochNumbers: epochNumbers?.split(",").map(Number),
      poolIds: poolIds?.split(","),
      includeAPR,
    });

    return new NextResponse(JSON.stringify(campaigns), {
      status: 200,
      headers: CACHE_HEADERS,
    });
  } catch (e) {
    console.error("Error in /api/campaigns:", e);
    return new NextResponse(JSON.stringify({message: (e as Error).message}), {
      status: 500,
      headers: CACHE_HEADERS,
    });
  }
}
