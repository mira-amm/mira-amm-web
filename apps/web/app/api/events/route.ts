/**
 * @api {get} /events Get all available events from one block to another (including all in between)
 */
import {NextRequest, NextResponse} from "next/server";
import {request, gql} from "graphql-request";
import {SQDIndexerUrl} from "../../../../../libs/web/src/utils/constants";
import {
  GeckoTerminalQueryResponses,
  SQDIndexerResponses,
} from "@/web/shared/types";
import {
  GeckoTerminalTypes,
  SQDIndexerTypes,
} from "../../../../../libs/web/shared/constants";
import {formatUnits} from "fuels";

interface FetchEventsParams {
  fromBlock: number;
  toBlock: number;
}

// Helper function to fetch and process events data for a block range
async function fetchEventsForBlockRange({
  fromBlock,
  toBlock,
}: FetchEventsParams): Promise<SQDIndexerResponses.Actions> {
  const query = gql`
    query GetActions($fromBlock: Int!, $toBlock: Int!) {
      actions(where: {blockNumber_gt: $fromBlock, blockNumber_lt: $toBlock}) {
        pool {
          id
        }
        asset0 {
          id
          decimals
        }
        asset1 {
          id
          decimals
        }
        amount1Out
        amount1In
        amount0Out
        amount0In
        reserves0After
        reserves1After
        type
        transaction
        recipient
        timestamp
        blockNumber
      }
    }
  `;

  try {
    // Send the GraphQL request to fetch actions
    const actionsData = await request<SQDIndexerResponses.Actions>({
      url: SQDIndexerUrl,
      document: query,
      variables: {fromBlock, toBlock},
    });

    return actionsData;
  } catch (error) {
    console.error(
      `Error fetching events for block range ${fromBlock} - ${toBlock}:`,
      error,
    );
    throw new Error(
      `Error fetching events for block range ${fromBlock} - ${toBlock}:`,
    );
  }
}

function createEventDataForJoinExitEvent(
  action: SQDIndexerResponses.Action,
  actionType:
    | SQDIndexerTypes.ActionTypes.JOIN
    | SQDIndexerTypes.ActionTypes.EXIT,
): GeckoTerminalQueryResponses.JoinExitEvent {
  const asset0Decimals = action.asset0.decimals;
  const asset1Decimals = action.asset1.decimals;
  //decimalizing based on asset's decimals

  const decimalizedReserves0After = formatUnits(
    action.reserves0After,
    asset0Decimals,
  );
  const decimalizedReserves1After = formatUnits(
    action.reserves1After,
    asset1Decimals,
  );
  let _amount0: number | string;
  let _amount1: number | string;

  //calculating amount0
  if (action.amount0In && action.amount0In != "0") {
    _amount0 = action.amount0In;
  } else if (action.amount0Out && action.amount0Out != "0") {
    _amount0 = action.amount0Out;
  } else {
    throw new Error(`Invalid swap event data: ${JSON.stringify(action)}`);
  }

  // calculation amount1
  if (action.amount1In && action.amount1In != "0") {
    _amount1 = action.amount1In;
  } else if (action.amount1Out && action.amount1Out != "0") {
    _amount1 = action.amount1Out;
  } else {
    throw new Error(`Invalid swap event data: ${JSON.stringify(action)}`);
  }

  //decimalizing based on asset's decimals
  const decimalizedAmount0 = formatUnits(_amount0, asset0Decimals);
  const decimalizedAmount1 = formatUnits(_amount1, asset1Decimals);

  const eventType =
    actionType == SQDIndexerTypes.ActionTypes.JOIN
      ? GeckoTerminalTypes.EventTypes.JOIN
      : GeckoTerminalTypes.EventTypes.EXIT;

  return {
    eventType: eventType,
    txnId: action.transaction,
    txnIndex: 0,
    eventIndex: 0,
    maker: action.recipient,
    pairId: action.pool.id,
    reserves: {
      asset0: decimalizedReserves0After,
      asset1: decimalizedReserves1After,
    },
    amount0: decimalizedAmount0,
    amount1: decimalizedAmount1,
  };
}

function createEventDataForSwapEvent(
  action: SQDIndexerResponses.Action,
): GeckoTerminalQueryResponses.SwapEvent {
  const asset0Decimals = action.asset0.decimals;
  const asset1Decimals = action.asset1.decimals;

  const decimalizedReserves0After = formatUnits(
    action.reserves0After,
    asset0Decimals,
  );
  const decimalizedReserves1After = formatUnits(
    action.reserves1After,
    asset1Decimals,
  );
  let event = {
    eventType: "swap",
    txnId: action.transaction,
    txnIndex: 0,
    eventIndex: 0,
    maker: action.recipient,
    pairId: action.pool.id,
    reserves: {
      asset0: decimalizedReserves0After,
      asset1: decimalizedReserves1After,
    },
  } as GeckoTerminalQueryResponses.SwapEvent;

  // not adding "0" value to the output as per the Gecko Terminal spec( only one pair should be present at any given)
  if (
    action.amount0In &&
    action.amount0In != "0" &&
    action.amount1Out &&
    action.amount1Out != "0"
  ) {
    event = {
      ...event,
      asset0In: formatUnits(action.amount0In, asset0Decimals),
      asset1Out: formatUnits(action.amount1Out, asset1Decimals),
      priceNative:
        parseFloat(formatUnits(action.amount0In, asset0Decimals)) /
        parseFloat(formatUnits(action.amount1Out, asset1Decimals)),
    };
  } else if (
    action.amount1In &&
    action.amount1In != "0" &&
    action.amount0Out &&
    action.amount0Out != "0"
  ) {
    event = {
      ...event,
      asset1In: formatUnits(action.amount1In, asset1Decimals),
      asset0Out: formatUnits(action.amount0Out, asset0Decimals),
      priceNative:
        parseFloat(formatUnits(action.amount1In, asset1Decimals)) /
        parseFloat(formatUnits(action.amount0Out, asset0Decimals)),
    };
  } else {
    throw new Error(`Invalid swap event data: ${JSON.stringify(action)}`);
  }
  return event;
}

// Handle GET requests for /api/events
export async function GET(req: NextRequest) {
  try {
    // Extract query parameters from the request URL
    const url = new URL(req.url);
    const fromBlock = parseInt(url.searchParams.get("fromBlock") || "0", 10);
    const toBlock = parseInt(url.searchParams.get("toBlock") || "0", 10);

    // Return a 400 error if fromBlock or toBlock is not provided
    if (!fromBlock || !toBlock) {
      return NextResponse.json(
        {error: "Both 'fromBlock' and 'toBlock' are required"},
        {status: 400},
      );
    }

    // Fetch events data for the given block range
    const actionsData = await fetchEventsForBlockRange({fromBlock, toBlock});
    // If no actions are found, return empty events list
    if (actionsData.actions.length === 0) {
      return NextResponse.json({events: []});
    }

    // Process the fetched events and map them to the expected format
    const events = actionsData.actions
      .map((action: SQDIndexerResponses.Action) => {
        // Create block from actions object
        const block: GeckoTerminalQueryResponses.Block = {
          blockNumber: action.blockNumber,
          blockTimestamp: action.timestamp,
        };

        // Identify event type
        const isJoinExitEvent =
          action.type === SQDIndexerTypes.ActionTypes.JOIN ||
          action.type === SQDIndexerTypes.ActionTypes.EXIT;
        const isSwapEvent = action.type === SQDIndexerTypes.ActionTypes.SWAP;
        let eventData:
          | GeckoTerminalQueryResponses.JoinExitEvent
          | GeckoTerminalQueryResponses.SwapEvent
          | null = null;
        // Extracting necessary data from action and creating respective event(join/swap)
        if (isJoinExitEvent) {
          eventData = createEventDataForJoinExitEvent(
            action,
            action.type as
              | SQDIndexerTypes.ActionTypes.JOIN
              | SQDIndexerTypes.ActionTypes.EXIT,
          );
        } else if (isSwapEvent) {
          eventData = createEventDataForSwapEvent(action);
        }

        // Return null when eventData is not created
        if (!eventData) {
          console.warn(`Unknown event type: ${action.type}. Possible reasons: 
              1. Malformed data (e.g., missing properties: ${JSON.stringify(action)}) 
              2. Unexpected event type (action.type is not 'JOIN or EXIT' or 'SWAP')
              3. Issue with event creation functions.`);
          return null;
        }

        return {
          block,
          ...eventData,
        };
      })
      // filtering out null items in the events list
      .filter((event): event is Exclude<typeof event, null> => event !== null);

    const eventsResponse: GeckoTerminalQueryResponses.EventsResponse = {
      events,
    };

    // Return the formatted events data
    return NextResponse.json(eventsResponse);
  } catch (error) {
    console.error("Error fetching events:", error);
    return NextResponse.json(
      {error: "Failed to fetch events data"},
      {status: 500},
    );
  }
}
