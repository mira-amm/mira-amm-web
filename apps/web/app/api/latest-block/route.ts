/**
 * @api {get} /latest-block Get latest block data (only if events data available for the block)
 */
// library
import {NextRequest, NextResponse} from "next/server";
import {request, gql} from "graphql-request";
import {DateTime} from "fuels";

// local imports
import {
  SQDIndexerUrl,
  NetworkUrl,
} from "../../../../../libs/web/src/utils/constants";
import {
  FuelAPIResponses,
  GeckoTerminalQueryResponses,
  SQDIndexerResponses,
} from "@/web/shared/types";

// Function to fetch the squid status (height)
async function fetchSquidStatus(
  url: string,
): Promise<SQDIndexerResponses.SquidStatus> {
  const query = gql`
    query GetSquidStatus {
      squidStatus {
        finalizedHeight
      }
    }
  `;

  try {
    const response = await request<{
      squidStatus: SQDIndexerResponses.SquidStatus;
    }>({
      url,
      document: query,
    });

    return response.squidStatus;
  } catch (error) {
    console.error("Error fetching squid status:", error);
    throw error;
  }
}

// Function to fetch the block data by height (timestamp)
async function fetchBlockByHeight(
  url: string,
  height: string,
): Promise<FuelAPIResponses.BlockByHeight> {
  const query = gql`
    query GetBlockByHeight($height: String!) {
      block(height: $height) {
        header {
          time
        }
      }
    }
  `;

  try {
    const response = await request<{block: FuelAPIResponses.BlockByHeight}>({
      url,
      document: query,
      variables: {height},
    });

    return response.block;
  } catch (error) {
    console.error("Error fetching block by height:", error);
    throw error;
  }
}

// Handle GET requests for /api/latest-block
export async function GET(req: NextRequest) {
  try {
    // Fetch the squid status to get the current height (latest block)
    const squidStatus = await fetchSquidStatus(SQDIndexerUrl);

    const blockNumber = squidStatus.finalizedHeight; // Latest block number

    // Fetch the block data (timestamp) using the block number
    const blockData = await fetchBlockByHeight(
      NetworkUrl,
      blockNumber.toString(),
    );

    const tai64TimeString = blockData.header.time.toString();

    // converting TAI64 string to normal datetime
    const blockDateTime: DateTime = DateTime.fromTai64(tai64TimeString);
    const blockTimeStampInSeconds = blockDateTime.toUnixSeconds();

    // Combine block number (height) and block timestamp
    const block: GeckoTerminalQueryResponses.Block = {
      blockNumber,
      blockTimestamp: blockTimeStampInSeconds,
    };

    const latestBlockResponse: GeckoTerminalQueryResponses.LatestBlockResponse =
      {
        block,
      };

    // Return the block data
    return NextResponse.json(latestBlockResponse);
  } catch (error) {
    console.error("error fetching latest block:", error);
    return NextResponse.json(
      {error: "An unexpected error occurred while fetching latest block"},
      {status: 500},
    );
  }
}
