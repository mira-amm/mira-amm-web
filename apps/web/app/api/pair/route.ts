/**
 * @api {get} /pair Get pair details by id
 */
import {NextRequest, NextResponse} from "next/server";
import {request, gql} from "graphql-request";
import {SQDIndexerUrl} from "../../../../../libs/web/src/utils/constants";
import {
  SQDIndexerResponses,
  GeckoTerminalQueryResponses,
} from "../../../../../libs/web/shared/types";
import {NotFoundError} from "../../../../../libs/web/src/utils/errors";

// Function to fetch pool details by ID
async function fetchPoolById(
  poolId: string,
): Promise<SQDIndexerResponses.Pool> {
  // Define the GraphQL query to fetch pool details by the given id
  const query = gql`
    query GetPoolById($id: String!) {
      poolById(id: $id) {
        id
        asset0 {
          id
        }
        asset1 {
          id
        }
        creationBlock
        creationTime
        creationTx
      }
    }
  `;

  // Send the GraphQL request to the indexer server to fetch the pool data
  const response = await request<{
    poolById: SQDIndexerResponses.Pool;
  }>({
    url: SQDIndexerUrl,
    document: query,
    variables: {id: poolId},
  });

  // throw error if pool is not found
  if (!response.poolById) {
    throw new NotFoundError(`Pool with ID: ${poolId} not found`);
  }

  // return pool/pair data
  return response.poolById;
}

function createPairFromPool(
  pool: SQDIndexerResponses.Pool,
): GeckoTerminalQueryResponses.Pair {
  const pair: GeckoTerminalQueryResponses.Pair = {
    id: pool.id,
    dexKey: "mira",
    asset0Id: pool.asset0.id,
    asset1Id: pool.asset1.id,
    createdAtBlockNumber: pool.creationBlock,
    createdAtBlockTimestamp: pool.creationTime,
    createdAtTxnId: pool.creationTx,
  };

  return pair;
}

// Handle GET requests for /api/pair
export async function GET(req: NextRequest) {
  try {
    const url = new URL(req.url);
    const poolId = url.searchParams.get("id");

    // Return a 400 error if no 'id' is provided
    if (!poolId) {
      return NextResponse.json(
        {error: "Pool ID(param: id) is required"},
        {status: 400},
      );
    }

    // Fetch the pool details
    const pool = await fetchPoolById(poolId);
    // Format pool data (pair and pool are synonymous) according to Gecko spec
    const pair = createPairFromPool(pool);
    const pairResponse: GeckoTerminalQueryResponses.PairResponse = {
      pair,
    };
    // Return the fetched pool(pair) data
    return NextResponse.json(pairResponse);
  } catch (error) {
    console.error("Error fetching pair data:", error);

    if (error instanceof NotFoundError) {
      return NextResponse.json({error: error.message}, {status: 404});
    }

    return NextResponse.json(
      {error: "An unexpected error occurred while fetching pair data"},
      {status: 500},
    );
  }
}
