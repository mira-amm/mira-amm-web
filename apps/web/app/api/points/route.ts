import { NextRequest, NextResponse } from "next/server";
import path from "path";
import { JSONEpochConfigService } from "@/src/models/campaigns/JSONEpochConfigService";
import { FileCachedPointsPerUserService } from "@/src/models/points/Points";
import { SENTIO_POINTS_ENDPOINT } from "@/src/utils/constants";

const CACHE_DURATION = 60;
const CACHE_STALE_WHILE_REVALIDATE = 60;
const cacheControlHeaders = {
  "Cache-Control": `public, s-maxage=${CACHE_DURATION}, stale-while-revalidate=${CACHE_STALE_WHILE_REVALIDATE}`,
  "Content-Type": "application/json",
};

const epochConfigService = new JSONEpochConfigService(
  path.join(process.cwd(), "../../libs/web/src", "models", "campaigns.json")
);

const pointsService = process.env.SENTIO_API_KEY
  ? new FileCachedPointsPerUserService(
      process.env.SENTIO_API_KEY,
      SENTIO_POINTS_ENDPOINT,
      epochConfigService
    )
  : null;

export async function GET(request: NextRequest): Promise<NextResponse> {
  if (!pointsService) {
    return new NextResponse(
      JSON.stringify({ message: "SENTIO_API_KEY and SENTIO_API_URL must be set" }),
      { status: 500, headers: cacheControlHeaders }
    );
  }

  const searchParams = request.nextUrl.searchParams;
  const address = searchParams.get("address") || undefined;
  const limit = searchParams.get("limit");
  const offset = searchParams.get("offset");

  const parsedLimit = limit ? Number(limit) : undefined;
  const parsedOffset = offset ? Number(offset) : undefined;

  try {
    const points = await pointsService.getPoints({
      address,
      limit: parsedLimit,
      offset: parsedOffset,
    });

    return new NextResponse(JSON.stringify(points), {
      status: 200,
      headers: cacheControlHeaders,
    });
  } catch (err) {
    console.error("Unhandled error in /api/points:", err);
    return new NextResponse(
      JSON.stringify({ message: (err as Error).message }),
      { status: 500, headers: cacheControlHeaders }
    );
  }
}
