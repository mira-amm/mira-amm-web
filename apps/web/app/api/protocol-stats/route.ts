import {NextResponse} from "next/server";
import {statsService} from "@/web/src/core/stats-service";
import type {ProtocolStatsData} from "@/web/src/types/protocol-stats";

/**
 * API route for fetching protocol statistics with 5-minute server-side caching
 * Uses existing fetchProtocolStats logic with simple cache control headers
 */
export async function GET(): Promise<NextResponse> {
  try {
    // Fetch protocol stats using existing logic
    const stats: ProtocolStatsData = await statsService.fetchProtocolStats();

    // Create response with protocol stats data
    const response = NextResponse.json(stats, {
      status: 200,
      headers: {
        // 5-minute server-side caching as specified in requirements
        "Cache-Control": "public, s-maxage=300, stale-while-revalidate=600",
        "CDN-Cache-Control": "public, s-maxage=300",
        "Vercel-CDN-Cache-Control": "public, s-maxage=300",
        // CORS headers for cross-origin requests
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET",
        "Access-Control-Allow-Headers": "Content-Type",
      },
    });

    return response;
  } catch (error) {
    console.error("Protocol stats API error:", error);

    // Return error response with appropriate status code
    return NextResponse.json(
      {
        error: "Failed to fetch protocol statistics",
        message:
          error instanceof Error ? error.message : "Unknown error occurred",
      },
      {
        status: 500,
        headers: {
          // No caching for error responses to allow quick retry
          "Cache-Control": "no-cache, no-store, must-revalidate",
          "Access-Control-Allow-Origin": "*",
          "Access-Control-Allow-Methods": "GET",
          "Access-Control-Allow-Headers": "Content-Type",
        },
      }
    );
  }
}

/**
 * Handle OPTIONS requests for CORS preflight
 */
export async function OPTIONS(): Promise<NextResponse> {
  return new NextResponse(null, {
    status: 200,
    headers: {
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Methods": "GET, OPTIONS",
      "Access-Control-Allow-Headers": "Content-Type",
    },
  });
}
