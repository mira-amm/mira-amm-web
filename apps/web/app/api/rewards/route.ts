/**
 * @api {get} /api/rewards Get rewards for a given user and epoch based on their LP tokens
 */
import {JSONEpochConfigService} from "@/src/models/campaigns/JSONEpochConfigService";
import {SentioJSONUserRewardsService} from "@/src/models/rewards/UserRewards";
import {NotFoundError} from "@/src/utils/errors";
import {NextRequest, NextResponse} from "next/server";
import path from "path";

const CACHE_HEADERS = {
  "Content-Type": "application/json",
  "Cache-Control": "public, max-age=3600, stale-while-revalidate=1800",
};

function createErrorResponse(status: number, message: string | object) {
  return new NextResponse(
    JSON.stringify(typeof message === "string" ? {message} : message),
    {
      status,
      headers: CACHE_HEADERS,
    }
  );
}

// Example query
// Key consideration, poolId does not include the hex prefix '0x'
// curl --request GET \
//   --url 'http://localhost:3000/api/rewards/?\
//  poolIds=286c479da40dc953bddc3bb4c453b608bba2e0ac483b077bd475174115395e6b-f8f8b6283d7fa5b672b530cbb84fcccb4ff8dc40f8176ef4544ddb1f1952ad07-false&\
//  epochNumbers=1&\
//  userId=0x69e6223f2adf576dfefb21873b78e31ba228b094d05f74f59ea60cbd1bf87d0d'

// poolIds and epochNumbers are comma separated,
// eg: epochNumbers=1,2,3&...
export async function GET(request: NextRequest) {
  if (!process.env.SENTIO_API_KEY || !process.env.SENTIO_API_URL) {
    return createErrorResponse(
      500,
      "Sentio API environment variables are not set."
    );
  }

  const searchParams = request.nextUrl.searchParams;
  const userId = searchParams.get("userId");
  const epochNumbersParam = searchParams.get("epochNumbers");
  const poolIdsParam = searchParams.get("poolIds");

  const missingParams = [];
  if (!userId) missingParams.push("userId");
  if (!epochNumbersParam) missingParams.push("epochNumbers");
  if (!poolIdsParam) missingParams.push("poolIds");

  if (missingParams.length > 0) {
    return createErrorResponse(
      400,
      `Missing required parameters: ${missingParams.join(", ")}`
    );
  }

  const epochNumbers = epochNumbersParam!.split(",").map(Number);
  const poolIds = poolIdsParam!.split(",");

  const userRewardsService = new SentioJSONUserRewardsService(
    process.env.SENTIO_API_URL,
    process.env.SENTIO_API_KEY,
    new JSONEpochConfigService(
      path.join(process.cwd(), "../../libs/web/src", "models", "campaigns.json")
    )
  );

  try {
    const rewards = await userRewardsService.getRewards({
      userId,
      epochNumbers,
      poolIds,
    });

    return new NextResponse(JSON.stringify(rewards), {
      status: 200,
      headers: CACHE_HEADERS,
    });
  } catch (e) {
    if (e instanceof NotFoundError) {
      return new NextResponse(JSON.stringify({}), {
        status: 200,
        headers: CACHE_HEADERS,
      });
    }

    console.error("Unhandled error in /api/rewards:", e);
    return createErrorResponse(500, (e as Error).message);
  }
}
