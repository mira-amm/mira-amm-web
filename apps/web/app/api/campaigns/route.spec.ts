import { describe, it, test, expect, beforeEach, afterEach, vi } from 'vitest'
import { NextRequest } from 'next/server'
import path from 'path'

const mockGetCampaigns = vi.fn()

vi.mock('@/src/models/campaigns/Campaign', () => {
  return {
    SentioJSONCampaignService: class {
      constructor(public apiUrl: string, public apiKey: string, public epochConfigService: unknown) {}
      getCampaigns = mockGetCampaigns
    },
  }
})

vi.mock('@/src/models/campaigns/JSONEpochConfigService', () => {
  return {
    JSONEpochConfigService: class {
      constructor(public configPath: string) {}
    },
  }
})

vi.stubEnv('SENTIO_API_KEY', 'fake-key')
vi.stubEnv('SENTIO_API_URL', 'https://fake.api')

import { GET } from './route'

const EXPECTED_CACHE_CONTROL = 'public, max-age=300, stale-while-revalidate=150'
const EXPECTED_CONTENT_TYPE = 'application/json'

describe('GET /api/campaigns', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })
  afterEach(() => {
    vi.restoreAllMocks()
  })

  it('instantiates JSONEpochConfigService with the correct path', () => {
    const { JSONEpochConfigService } = require('@/src/models/campaigns/JSONEpochConfigService')
    expect(JSONEpochConfigService).toHaveBeenCalledOnce()
    const expectedPath = path.join(
      process.cwd(),
      '../../libs/web/src',
      'models',
      'campaigns.json'
    )
    expect(JSONEpochConfigService).toHaveBeenCalledWith(expectedPath)
  })

  it('returns 200 and calls getCampaigns() with no filters when no query params', async () => {
    const dummyCampaigns = [{ foo: 'bar' }]
    mockGetCampaigns.mockResolvedValueOnce(dummyCampaigns)

    const req = new NextRequest('https://example.com/api/campaigns')
    const res = await GET(req)

    expect(mockGetCampaigns).toHaveBeenCalledOnce()
    expect(mockGetCampaigns).toHaveBeenCalledWith({
      epochNumbers: undefined,
      poolIds: undefined,
      includeAPR: false,
    })

    expect(res.status).toBe(200)
    expect(res.headers.get('Content-Type')).toBe(EXPECTED_CONTENT_TYPE)
    expect(res.headers.get('Cache-Control')).toBe(EXPECTED_CACHE_CONTROL)
    expect(await res.json()).toEqual(dummyCampaigns)
  })

  test.each([
    // [ queryString, expectedArguments ]
    ['epochNumbers=1,2', { epochNumbers: [1, 2], poolIds: undefined, includeAPR: false }],
    ['poolIds=a,b',      { epochNumbers: undefined, poolIds: ['a', 'b'], includeAPR: false }],
    ['includeAPR=true',  { epochNumbers: undefined, poolIds: undefined, includeAPR: true }],
    ['epochNumbers=3,4&poolIds=x,y&includeAPR=true', {
      epochNumbers: [3, 4],
      poolIds: ['x', 'y'],
      includeAPR: true,
    }],
  ])(
    'parses query "%s" correctly',
    async (queryString, expectedArgs) => {
      mockGetCampaigns.mockResolvedValueOnce([])

      const req = new NextRequest(`https://example.com/api/campaigns?${queryString}`)
      const res = await GET(req)

      expect(mockGetCampaigns).toHaveBeenCalledOnce()
      expect(mockGetCampaigns).toHaveBeenCalledWith(expectedArgs)
      expect(res.status).toBe(200)
      expect(await res.json()).toEqual([])
    }
  )

  it('returns 500 and error message when getCampaigns throws', async () => {
    mockGetCampaigns.mockRejectedValueOnce(new Error('service down'))

    const req = new NextRequest('https://example.com/api/campaigns')
    const res = await GET(req)

    expect(res.status).toBe(500)
    expect(res.headers.get('Content-Type')).toBe(EXPECTED_CONTENT_TYPE)
    expect(await res.json()).toEqual({ message: 'service down' })
  })

  it('response JSON matches snapshot', async () => {
    const sample = [
      {
        epoch: { number: 7, startDate: '2025-06-01T00:00:00.000Z', endDate: '2025-06-08T00:00:00.000Z' },
        pool: { id: 'pool-id', symbols: ['USDC', 'FUEL'], lpToken: '0xabc' },
        rewards: [{ dailyAmount: 123, assetId: '0xfee' }],
        status: 'inprogress' as const,
      },
    ]
    mockGetCampaigns.mockResolvedValueOnce(sample)

    const req = new NextRequest('https://example.com/api/campaigns')
    const res = await GET(req)
    const json = await res.json()

    expect(json).toMatchSnapshot()
  })

  it('waits for async getCampaigns (fake timers)', async () => {
    const delayedResult = [{ dummy: true }]
    mockGetCampaigns.mockImplementation(
      () => new Promise((resolve) => setTimeout(() => resolve(delayedResult), 500))
    )

    const req = new NextRequest('https://example.com/api/campaigns')
    vi.useFakeTimers()
    const promise = GET(req)
    vi.advanceTimersByTime(500)
    const res = await promise
    expect(res.status).toBe(200)
    expect(await res.json()).toEqual(delayedResult)
    vi.useRealTimers()
  })
})
