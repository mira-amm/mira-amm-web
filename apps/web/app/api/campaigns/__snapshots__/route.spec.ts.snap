// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GET /api/campaigns > response JSON matches snapshot 1`] = `
[
  {
    "epoch": {
      "endDate": "2025-06-08T00:00:00.000Z",
      "number": 7,
      "startDate": "2025-06-01T00:00:00.000Z",
    },
    "pool": {
      "id": "pool-id",
      "lpToken": "0xabc",
      "symbols": [
        "USDC",
        "FUEL",
      ],
    },
    "rewards": [
      {
        "assetId": "0xfee",
        "dailyAmount": 123,
      },
    ],
    "status": "inprogress",
  },
]
`;
